### ISSUE TITLE
Run Travis test against custom compiled postgres

### ISSUE THREAD

**Issue by anarazel** [citusdata/citus#1034]
That will:
- allow to run isolationtester tests as part of the normal tests (#917)
- allow testing against assertions (#208)
- allow running standard postgres tests as part of our tests (#1017)
- in theory it'd allow to run valgrind tests, but they take too long for travis (3-5h)

Possibly more.

@jasonmp85 ^

**Comment by jasonmp85** [comment]
We discussed whether to omit `ruleutils` files _et al_ from coverage reports. @anarazel thought they should be included if we run standard PostgreSQL tests as part of our tests; until then, I've omitted them (as we did in e.g. `pg_shard`) as part of #1035.

Whoever tackles #1017 above might consider re-including them.

**Comment by anarazel** [comment]
> We discussed whether to omit `ruleutils` files _et al_ from coverage
> reports. @anarazel thought they should be included if we run standard
> PostgreSQL tests as part of our tests; until then, I've omitted them
> (as we did in e.g. `pg_shard`) as part of #1035.

Hm? No, I didn't.  I want to include the coverage information from
postgres tests against citus in the coverage information, but I'm fine
with omitting ruleutils/readfuncs from all.

**Comment by jasonmp85** [comment]
Ah, then I misunderstood 😄. Either way, they're omitted now, since ≈ 50% of the lines are not hit by our tests (and the large line counts mean that has a large impact), and the `ruleutils` code is already thoroughly tested by PostgreSQL.

**Comment by byucesoy** [comment]
I will implement this functionality and before starting I just want to confirm that my understanding of the issue is correct.

Basically we want to run push tests against a PostgreSQL which is compiled from the source(not pre-compiled and installed via sudo apt-get install). Doing this will allow us to run isolation tests, assert enabled tests etc. as part of travis.

To add this functionality, I will need;
* modify travis.yml file to download, extract and compile PostgreSQL source code.
    * For this operation I will probably add some new scripts to [tools/travis](https://github.com/citusdata/tools/tree/develop/travis) and use them in travis.yml
    * I will also compile "postgrsql_src/test/isolation", because this is not compiled by default
    * I will use hard-coded URL for PostgreSQL's source location(?)

* add check-isolation and check-vanilla tests to check-full, so that they will be run by travis as part of push tests
    * Doing this will increase duration of push tests, are we OK with that?

Apart from that, check-isolation and check-vanilla tests crashes at the moment, probably a commit breaks them and we did not realized because isolation tests are not part of push tests. I will also fix those issues.

@anarazel, @jasonmp85, is there anything I missed? If not, I will start implementing this Wednesday morning Turkey Time.

**Comment by anarazel** [comment]
> modify travis.yml file to download, extract and compile PostgreSQL source code. 

Travis supports some caching of build output.  Might be worthwhile to check how much that saves.


> I will use hard-coded URL for PostgreSQL's source location(?)

I'd assume you do a git checkout?


> add check-isolation and check-vanilla tests to check-full, so that they will be run by travis as part of push tests 
That's probably not the greatest idea, because it'll also fail the tests for all people locally developing against an pre-built PG.  How about adding those tests to travis instead.

> Apart from those, check-isolation and check-vanilla tests crashes at the moment, probably a commit breaks them and we did not realized because those tests are not part of the push tests. I will also fix those issues.

:( that's not good...

> Doing this will increase duration of push tests, are we OK with that?

We probably don't want to do that every single build, because testing against a release build is also useful.   Perhaps we could rig things so that the "hypothetical merge to master" commits get tested against release, and the branch heads against this?

> @anarazel, @jasonmp85, is there anything I missed? If not, I will start implementing this Wednesday morning Turkey Time.

Except for the minor bits above, it sounds good.

**Comment by jasonmp85** [comment]
This got me off trying to use `quilt` to patch the PGDG packages to enable assertions. Seemed to work pretty well.

PostgreSQL already installs `pg_regress`, so the only missing pieces (I think?) are the `regress.so` and test files within `src/test/regress`. Should be easy enough to add them to a modified `deb`.

I haven't had time to look at the isolation tester at the moment, but I'm hopeful the same can be done there.

**Comment by jasonmp85** [comment]
Don't change anything due to the above, I was just curious about whether we could avoid having to rebuild PostgreSQL each time (Travis [directory caching](https://docs.travis-ci.com/user/caching/) might be a better short-term solution).
